File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/BleMdsClient.kt
package com.polar.androidcommunications.api.ble.model.gatt.client

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.exceptions.BleAttributeError
import com.polar.androidcommunications.api.ble.exceptions.BleDisconnected
import com.polar.androidcommunications.api.ble.exceptions.BleTimeout
import com.polar.androidcommunications.api.ble.model.gatt.BleGattBase
import com.polar.androidcommunications.api.ble.model.gatt.BleGattTxInterface
import com.polar.androidcommunications.common.ble.AtomicSet
import com.polar.androidcommunications.common.ble.ChannelUtils
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.util.UUID
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

/**
 * The `BleMdsClient` class implements BLE MDS (Memfault Diagnostic Service) client for receiving
 * telemetry diagnostics data from a BLE MDS service.
 *
 * Service flow:
 * 1. After connection [MDS_SUPPORTED_FEATURES], [MDS_DEVICE_IDENTIFIER], [MDS_DATA_URI] and
 *    [MDS_AUTHORIZATION] are auto-read, and [MDS_DATA_EXPORT] notifications are auto-enabled.
 * 2. Call [startMdsNotifications] to register an observer and enable streaming via
 *    [MDS_DATA_EXPORT].
 * 3. Observe [TelemetryConfiguration] for continuous streaming; each
 *    emission carries the latest device context together with the export data chunk.
 *
 * @property txInterface The BLE GATT transmitter interface used to send and receive GATT requests
 * and responses.
 */
class BleMdsClient(txInterface: BleGattTxInterface) : BleGattBase(txInterface, MDS_SERVICE) {

    companion object {
        private const val TAG = "BleMdsClient"
        val MDS_SERVICE: UUID = UUID.fromString("54220000-f6a5-4007-a371-722f4ebd8436")
        val MDS_SUPPORTED_FEATURES: UUID = UUID.fromString("54220001-f6a5-4007-a371-722f4ebd8436")
        val MDS_DEVICE_IDENTIFIER: UUID = UUID.fromString("54220002-f6a5-4007-a371-722f4ebd8436")
        val MDS_DATA_URI: UUID = UUID.fromString("54220003-f6a5-4007-a371-722f4ebd8436")
        val MDS_AUTHORIZATION: UUID = UUID.fromString("54220004-f6a5-4007-a371-722f4ebd8436")
        val MDS_DATA_EXPORT: UUID = UUID.fromString("54220005-f6a5-4007-a371-722f4ebd8436")
        val MDS_CONFIGURATOR_DESCRIPTOR: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    }

    /**
     * Container that carries the latest MDS device context together with a single chunk of
     * streaming export data received from the [MDS_DATA_EXPORT] notification.
     *
     * All context fields are populated from auto-reads performed after service discovery and are
     * `null` until their respective reads complete.
     *
     * [deviceIdentifier] — unique identifier of the device (from [MDS_DEVICE_IDENTIFIER]).
     *
     * [dataUri] — Memfault data ingestion URI (from [MDS_DATA_URI]).
     *
     * [authorization] — HTTP authorization header value to include when forwarding data to
     * Memfault, e.g. `"Memfault-Project-Key:YOUR_PROJECT_KEY"` (from [MDS_AUTHORIZATION]).
     *
     * [exportData] — raw bytes of one [MDS_DATA_EXPORT] notification frame.
     */
    data class TelemetryConfiguration(
        val deviceIdentifier: String?,
        val dataUri: String?,
        val authorization: String?,
        val exportData: ByteArray,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as TelemetryConfiguration
            return deviceIdentifier == other.deviceIdentifier &&
                    dataUri == other.dataUri &&
                    authorization == other.authorization &&
                    exportData.contentEquals(other.exportData)
        }

        override fun hashCode(): Int {
            var result = deviceIdentifier?.hashCode() ?: 0
            result = 31 * result + (dataUri?.hashCode() ?: 0)
            result = 31 * result + (authorization?.hashCode() ?: 0)
            result = 31 * result + exportData.contentHashCode()
            return result
        }
    }

    // Device context — populated as reads complete after service discovery.
    @Volatile internal var deviceIdentifier: String? = null
    @Volatile internal var dataUri: String? = null
    @Volatile internal var authorization: String? = null
    @Volatile internal var supportedFeatures: Int? = null

    // Modulo-32 sequence counter from the last accepted DATA_EXPORT notification (-1 = none yet).
    @Volatile private var lastSeqCounter: Int = -1

    private val mdsObserverAtomicList = AtomicSet<Channel<TelemetryConfiguration>>()
    private val supportedFeaturesReadLock = Any()
    @Volatile private var supportedFeaturesRead = CompletableDeferred<Int>()
    val mdsResponseQueue = LinkedBlockingQueue<Pair<ByteArray, Int>>()

    init {
        // Auto-read readable characteristics after service discovery.
        addCharacteristicRead(MDS_SUPPORTED_FEATURES)
        addCharacteristicRead(MDS_DEVICE_IDENTIFIER)
        addCharacteristicRead(MDS_DATA_URI)
        addCharacteristicRead(MDS_AUTHORIZATION)
        // Register DATA_EXPORT for auto-notification setup and write-callback routing.
        addCharacteristic(MDS_DATA_EXPORT, PROPERTY_NOTIFY or PROPERTY_WRITE or PROPERTY_WRITE_NO_RESPONSE)
    }

    override fun reset() {
        super.reset()
        deviceIdentifier = null
        dataUri = null
        authorization = null
        supportedFeatures = null
        synchronized(supportedFeaturesReadLock) {
            supportedFeaturesRead.completeExceptionally(BleDisconnected())
            supportedFeaturesRead = CompletableDeferred()
        }
        lastSeqCounter = -1
        mdsResponseQueue.clear()
        ChannelUtils.postDisconnectedAndClearList(mdsObserverAtomicList)
    }

    override suspend fun clientReady(checkConnection: Boolean) {
        val currentSupportedFeaturesRead = synchronized(supportedFeaturesReadLock) {
            supportedFeaturesRead
        }
        waitNotificationEnabled(MDS_DATA_EXPORT, checkConnection)
        currentSupportedFeaturesRead.await()
    }

    override fun processServiceData(
        characteristic: UUID,
        data: ByteArray,
        status: Int,
        notifying: Boolean
    ) {
        when (characteristic) {
            MDS_SUPPORTED_FEATURES -> {
                if (status == ATT_SUCCESS) {
                    val features = data.foldIndexed(0) { i, acc, byte -> acc or ((byte.toInt() and 0xFF) shl (8 * i)) }
                    supportedFeatures = features
                    synchronized(supportedFeaturesReadLock) {
                        supportedFeaturesRead.complete(features)
                    }
                    BleLogger.d(TAG, "MDS Supported Features received (${data.size} bytes): ${data.map { it }}")
                } else {
                    BleLogger.e(TAG, "MDS Supported Features read error: $status")
                    synchronized(supportedFeaturesReadLock) {
                        supportedFeaturesRead.completeExceptionally(
                            BleAttributeError("MDS Supported Features attribute ", status)
                        )
                    }
                }
            }
            MDS_DEVICE_IDENTIFIER -> {
                if (status == ATT_SUCCESS) {
                    deviceIdentifier = data.toString(Charsets.UTF_8)
                    BleLogger.d(TAG, "MDS Device Identifier received: $deviceIdentifier")
                } else {
                    BleLogger.e(TAG, "MDS Device Identifier read error: $status")
                }
            }
            MDS_DATA_URI -> {
                if (status == ATT_SUCCESS) {
                    dataUri = data.toString(Charsets.UTF_8)
                    BleLogger.d(TAG, "MDS Data URI received: $dataUri")
                } else {
                    BleLogger.e(TAG, "MDS Data URI read error: $status")
                }
            }
            MDS_AUTHORIZATION -> {
                if (status == ATT_SUCCESS) {
                    authorization = data.toString(Charsets.UTF_8)
                    BleLogger.d(TAG, "MDS Authorization received: $authorization")
                } else {
                    BleLogger.e(TAG, "MDS Authorization read error: $status")
                }
            }
            MDS_DATA_EXPORT -> {
                if (notifying) {
                    BleLogger.d(TAG, "MDS Data Export notification received (${data.size} bytes)")

                    if (data.isEmpty()) {
                        BleLogger.w(TAG, "MDS Data Export: empty notification, ignoring")
                        return
                    }

                    val seqCounter = data[0].toInt() and 0x1F
                    val payload = data.sliceArray(1 until data.size)

                    if (lastSeqCounter != -1) {
                        if (seqCounter == lastSeqCounter) {
                            BleLogger.d(TAG, "MDS Data Export: duplicate chunk seq=$seqCounter — dropped")
                            return
                        }
                        val dropped = (seqCounter - lastSeqCounter - 1 + 32) % 32
                        if (dropped > 0) {
                            BleLogger.w(TAG, "MDS Data Export: $dropped chunk(s) lost (seq $lastSeqCounter → $seqCounter)")
                        }
                    }
                    lastSeqCounter = seqCounter

                    ChannelUtils.emitNext(mdsObserverAtomicList) { observer ->
                        val notificationData = TelemetryConfiguration(deviceIdentifier, dataUri, authorization, payload)
                        observer.trySend(notificationData)
                    }
                }
            }
            else -> BleLogger.d(TAG, "processServiceData: unhandled characteristic $characteristic status: $status")
        }
    }

    override fun processServiceDataWritten(characteristic: UUID, status: Int) {
        BleLogger.d(TAG, "processServiceDataWritten characteristic: $characteristic status: $status")
        if (characteristic == MDS_DATA_EXPORT) {
            if (status != ATT_SUCCESS) {
                BleLogger.e(TAG, "MDS Data Export write failed with status: $status")
            }
            mdsResponseQueue.add(Pair(byteArrayOf(), status))
        }
    }

    fun startMdsNotifications(checkConnection: Boolean): Flow<TelemetryConfiguration> {
        if (checkConnection && !txInterface.isConnected()) {
            throw BleDisconnected()
        }

        lastSeqCounter = -1
        val observer = Channel<TelemetryConfiguration>(Channel.BUFFERED)
        mdsObserverAtomicList.add(observer)

        try {
            sendControlPointCommand(1)
        } catch (error: Throwable) {
            mdsObserverAtomicList.remove(observer)
            observer.close()
            throw error
        }

        return flow {
            try {
                for (notification in observer) {
                    emit(notification)
                }
            } finally {
                mdsObserverAtomicList.remove(observer)
                observer.close()
            }
        }
    }

    fun monitorMdsNotifications(checkConnection: Boolean): Flow<TelemetryConfiguration> {
        return ChannelUtils.monitorNotifications(mdsObserverAtomicList, txInterface, checkConnection)
    }

    fun sendControlPointCommand(command: Int): Int {
        mdsResponseQueue.clear()
        txInterface.transmitMessages(MDS_SERVICE, MDS_DATA_EXPORT, listOf(byteArrayOf(command.toByte())), true)
        val pair = mdsResponseQueue.poll(30, TimeUnit.SECONDS)
        if (pair != null) {
            if (pair.second == ATT_SUCCESS) {
                return 0
            } else {
                throw BleAttributeError("mds attribute ", pair.second)
            }
        }
        throw BleTimeout("Mds response failed to receive in timeline")
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/BleMdsClient.swift
import CoreBluetooth
import Foundation
import Combine

open class BleMdsClient: BleGattClientBase, @unchecked Sendable {
    
    /// NOTE: this is an experimental BLE client
    ///  intended for Polar internal use only. Polar will not support 3rd party users with this API.
    /// Connection parameters read from the Memfault Diagnostic Service (MDS) on the BLE device.
    ///
    /// Returned by `PolarDeviceTelemetryApi.getTelemetryConnectionInfo(_:)`.
    /// Forward `deviceIdentifier`, `dataUri` and `authorization` to the Memfault cloud API
    /// when uploading chunks received via `startMemfaultStreaming(_:)`.
    public struct DeviceTelemetryConfiguration: Equatable, Sendable {

        /// Memfault device key — used as the `{device_identifier}` path segment in cloud API calls.
        public let deviceIdentifier: String

        /// URL for the Memfault cloud chunk upload endpoint, e.g. `"https://nrf-chunks.memfault.com/api/v0/chunks/"`.
        /// Append the device identifier to complete the upload path: `<dataUri><deviceIdentifier>`
        public let dataUri: String

        /// Authorization token (project key or Bearer token). May be empty if the device does not
        /// require authorization. When non-empty, set as `Memfault-Project-Key` or `Authorization`
        /// HTTP header when uploading chunks.
        public let authorization: String

        /// Raw bitmask read from the MDS Supported Features characteristic (UUID 54220001).
        ///
        /// Each bit indicates a capability of the device's MDS implementation.
        /// Currently the spec defines only bit 0:
        /// - bit 0 (`0x01`) — live streaming of Memfault diagnostic chunks is supported.
        ///
        /// All other bits are reserved for future use.
        /// The value `0x00` means this is the baseline MDS profile with no optional features enabled.
        public let supportedFeatures: UInt32
        
        public init(deviceIdentifier: String,
                    dataUri: String,
                    authorization: String,
                    supportedFeatures: UInt32) {
            self.deviceIdentifier = deviceIdentifier
            self.dataUri = dataUri
            self.authorization = authorization
            self.supportedFeatures = supportedFeatures
        }
    }
    
    /// NOTE: this is an experimental type intended for Polar internal use only.
    /// Memfault MDS telemetry configuration with cloud upload parameters
    /// Used by the public DeviceTelemetryConfiguration enum
    public struct MemfaultTelemetryConfiguration: Equatable, Sendable {
        public let deviceIdentifier: String
        /// URI provided by device for uploading telemetry data via HTTP to Memfault service.
        public let dataUri: String?
        /// Authorization token (Memfault-Project-Key value) for HTTP header when uploading telemetry data.
        public let authorization: String?
        /// BLE service specific list of supported features
        public let supportedFeatures: [UInt32]
        
        public init(deviceIdentifier: String, dataUri: String?, authorization: String?, supportedFeatures: [UInt32]) {
            self.deviceIdentifier = deviceIdentifier
            self.dataUri = dataUri
            self.authorization = authorization
            self.supportedFeatures = supportedFeatures
        }
    }

    // MARK: - GATT UUIDs
    public static let MDS_SERVICE            = CBUUID(string: "54220000-f6a5-4007-a371-722f4ebd8436")
    public static let MDS_SUPPORTED_FEATURES = CBUUID(string: "54220001-f6a5-4007-a371-722f4ebd8436")
    public static let MDS_DEVICE_IDENTIFIER  = CBUUID(string: "54220002-f6a5-4007-a371-722f4ebd8436")
    public static let MDS_DATA_URI           = CBUUID(string: "54220003-f6a5-4007-a371-722f4ebd8436")
    public static let MDS_AUTHORIZATION      = CBUUID(string: "54220004-f6a5-4007-a371-722f4ebd8436")
    public static let MDS_DATA_EXPORT        = CBUUID(string: "54220005-f6a5-4007-a371-722f4ebd8436")

    // MARK: - Internal response continuations for read characteristics
    private var featuresContinuation: CheckedContinuation<Data, Error>? = nil
    private var deviceIdContinuation: CheckedContinuation<Data, Error>? = nil
    private var dataUriContinuation: CheckedContinuation<Data, Error>? = nil
    private var authContinuation: CheckedContinuation<Data, Error>? = nil

    // MARK: - Streaming (DATA_EXPORT notifications → Memfault chunks)
    private let chunkStreams = StreamContinuationList<Data>()

    // Continuation awaiting the ATT Write Response for a DATA_EXPORT mode write.
    // Resumed by serviceDataWritten(_:err:) when didWriteValueFor fires.
    private var writeModeContinuation: CheckedContinuation<Void, Error>? = nil

    // Continuation awaiting the ATT result of a CCCD enable/disable write for DATA_EXPORT.
    // Resumed by the notifyDescriptorWritten override below.
    private var cccdContinuation: CheckedContinuation<Void, Error>? = nil
    
    // Tracks the last valid sequence number (bits [4:0] of byte 0) to detect duplicates and gaps.
    // Sequence numbers are modulo-32 (0-31). Reset to nil on disconnect or stream start.
    private var lastSeqNumber: UInt8? = nil

    // MARK: - Init

    public init(gattServiceTransmitter: BleAttributeTransportProtocol) {
        super.init(serviceUuid: BleMdsClient.MDS_SERVICE,
                   gattServiceTransmitter: gattServiceTransmitter)
        // Readable characteristics — discovered and read on-demand via readValue()
        addCharacteristicRead(BleMdsClient.MDS_SUPPORTED_FEATURES)
        addCharacteristicRead(BleMdsClient.MDS_DEVICE_IDENTIFIER)
        addCharacteristicRead(BleMdsClient.MDS_DATA_URI)
        addCharacteristicRead(BleMdsClient.MDS_AUTHORIZATION)
        // DATA_EXPORT is notifiable; notification is enabled on-demand (not auto on connect)
        addCharacteristic(BleMdsClient.MDS_DATA_EXPORT)
    }

    // MARK: - BleGattClientBase overrides

    override public func disconnected() {
        super.disconnected()
        
        // Cancel any pending reads
        featuresContinuation?.resume(throwing: BleGattException.gattDisconnected)
        featuresContinuation = nil
        deviceIdContinuation?.resume(throwing: BleGattException.gattDisconnected)
        deviceIdContinuation = nil
        dataUriContinuation?.resume(throwing: BleGattException.gattDisconnected)
        dataUriContinuation = nil
        authContinuation?.resume(throwing: BleGattException.gattDisconnected)
        authContinuation = nil
        
        cccdContinuation?.resume(throwing: BleGattException.gattDisconnected)
        cccdContinuation = nil
        writeModeContinuation?.resume(throwing: BleGattException.gattDisconnected)
        writeModeContinuation = nil
        chunkStreams.finish(throwing: BleGattException.gattDisconnected)
        
        // Reset sequence tracking on disconnect
        lastSeqNumber = nil
    }

    override public func processServiceData(_ chr: CBUUID, data: Data, err: Int) {
        if chr == BleMdsClient.MDS_SUPPORTED_FEATURES {
            guard let continuation = featuresContinuation else { return }
            featuresContinuation = nil
            if err == 0 {
                BleLogger.trace("MDS_SUPPORTED_FEATURES received (\(data.count) bytes)")
                continuation.resume(returning: data)
            } else {
                BleLogger.error("MDS_SUPPORTED_FEATURES read error: \(err)")
                continuation.resume(throwing: BleGattException.gattCharacteristicError)
            }
        } else if chr == BleMdsClient.MDS_DEVICE_IDENTIFIER {
            guard let continuation = deviceIdContinuation else { return }
            deviceIdContinuation = nil
            if err == 0 {
                BleLogger.trace("MDS_DEVICE_IDENTIFIER received (\(data.count) bytes)")
                continuation.resume(returning: data)
            } else {
                BleLogger.error("MDS_DEVICE_IDENTIFIER read error: \(err)")
                continuation.resume(throwing: BleGattException.gattCharacteristicError)
            }
        } else if chr == BleMdsClient.MDS_DATA_URI {
            guard let continuation = dataUriContinuation else { return }
            dataUriContinuation = nil
            if err == 0 {
                BleLogger.trace("MDS_DATA_URI received (\(data.count) bytes)")
                continuation.resume(returning: data)
            } else {
                BleLogger.error("MDS_DATA_URI read error: \(err)")
                continuation.resume(throwing: BleGattException.gattCharacteristicError)
            }
        } else if chr == BleMdsClient.MDS_AUTHORIZATION {
            guard let continuation = authContinuation else { return }
            authContinuation = nil
            if err == 0 {
                BleLogger.trace("MDS_AUTHORIZATION received (\(data.count) bytes)")
                continuation.resume(returning: data)
            } else {
                BleLogger.error("MDS_AUTHORIZATION read error: \(err)")
                continuation.resume(throwing: BleGattException.gattCharacteristicError)
            }
        } else if chr == BleMdsClient.MDS_DATA_EXPORT {
            if err == 0 {
                BleLogger.trace("MDS_DATA_EXPORT chunk received (\(data.count) bytes)")
                // Validate sequence number and filter duplicates
                if validateAndFilterChunk(data) {
                    chunkStreams.yield(Data(data.dropFirst()))
                }
            } else {
                BleLogger.error("MDS_DATA_EXPORT notification error: \(err)")
                chunkStreams.finish(throwing: BleGattException.gattCharacteristicNotifyError(
                    errorCode: err,
                    errorDescription: "MDS_DATA_EXPORT notification error \(err)"
                ))
            }
        }
    }

    /// MDS is considered ready as soon as the service is discovered; notifications are on-demand.
    public override func clientReady(_ checkConnection: Bool) -> AnyPublisher<Never, Error> {
        waitDiscovered(checkConnection: checkConnection)
    }

    /// Intercept CCCD write results for MDS_DATA_EXPORT and forward them to `cccdContinuation`.
    override public func notifyDescriptorWritten(_ chr: CBUUID, enabled: Bool, err: Int) {
        super.notifyDescriptorWritten(chr, enabled: enabled, err: err)
        guard chr.isEqual(BleMdsClient.MDS_DATA_EXPORT) else { return }
        guard let continuation = cccdContinuation else { return }
        cccdContinuation = nil
        if err == 0 {
            continuation.resume()
        } else {
            continuation.resume(throwing: BleGattException.gattCharacteristicNotifyError(
                errorCode: err,
                errorDescription: "MDS_DATA_EXPORT CCCD write failed with err \(err)"
            ))
        }
    }

    /// Read Supported Features bitmask.
    /// Bit 0 = streaming supported, Bit 1 = data export supported.
    /// The MDS spec defines this as a variable-length little-endian field (1–4 bytes).
    open func readSupportedFeatures() async throws -> UInt32 {
        guard let transport = gattServiceTransmitter else {
            throw BleGattException.gattTransportNotAvailable
        }
        
        let data = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Data, Error>) in
            featuresContinuation = continuation
            do {
                try transport.readValue(self, serviceUuid: BleMdsClient.MDS_SERVICE,
                                        characteristicUuid: BleMdsClient.MDS_SUPPORTED_FEATURES)
            } catch {
                featuresContinuation = nil
                continuation.resume(throwing: error)
            }
        }
        
        guard data.count >= 1 && data.count <= 4 else {
            BleLogger.error("MDS_SUPPORTED_FEATURES unexpected length: \(data.count) bytes")
            throw BleGattException.gattOperationModeChange()
        }
        // Zero-pad to 4 bytes so we can always load as a little-endian UInt32.
        var padded = Data(repeating: 0, count: 4)
        padded.replaceSubrange(0..<data.count, with: data)
        return padded.withUnsafeBytes { $0.load(as: UInt32.self) }.littleEndian
    }

    /// Read the Memfault device identifier (UTF-8 string).
    open func readDeviceIdentifier() async throws -> String {
        guard let transport = gattServiceTransmitter else {
            throw BleGattException.gattTransportNotAvailable
        }
        
        let data = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Data, Error>) in
            deviceIdContinuation = continuation
            do {
                try transport.readValue(self, serviceUuid: BleMdsClient.MDS_SERVICE,
                                        characteristicUuid: BleMdsClient.MDS_DEVICE_IDENTIFIER)
            } catch {
                deviceIdContinuation = nil
                continuation.resume(throwing: error)
            }
        }
        
        return String(bytes: data, encoding: .utf8) ?? ""
    }

    /// Read the cloud data URI (UTF-8 string).
    open func readDataUri() async throws -> String {
        guard let transport = gattServiceTransmitter else {
            throw BleGattException.gattTransportNotAvailable
        }
        
        let data = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Data, Error>) in
            dataUriContinuation = continuation
            do {
                try transport.readValue(self, serviceUuid: BleMdsClient.MDS_SERVICE,
                                        characteristicUuid: BleMdsClient.MDS_DATA_URI)
            } catch {
                dataUriContinuation = nil
                continuation.resume(throwing: error)
            }
        }
        
        return String(bytes: data, encoding: .utf8) ?? ""
    }

    /// Read the authorization token (UTF-8 string, may be empty).
    open func readAuthorization() async throws -> String {
        guard let transport = gattServiceTransmitter else {
            throw BleGattException.gattTransportNotAvailable
        }
        
        let data = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Data, Error>) in
            authContinuation = continuation
            do {
                try transport.readValue(self, serviceUuid: BleMdsClient.MDS_SERVICE,
                                        characteristicUuid: BleMdsClient.MDS_AUTHORIZATION)
            } catch {
                authContinuation = nil
                continuation.resume(throwing: error)
            }
        }
        
        return String(bytes: data, encoding: .utf8) ?? ""
    }

    /// Returns a stream of raw Memfault chunk Data packets delivered via MDS_DATA_EXPORT notifications.
    /// Each emitted Data value is one Memfault chunk, ready to be forwarded to the Memfault cloud.
    /// The stream ends when the device disconnects, an error occurs, or streaming is stopped.
    open func observeMemfaultChunks(checkConnection: Bool) -> AsyncThrowingStream<Data, Error> {
        return chunkStreams.makeStream(transport: gattServiceTransmitter, checkConnection: checkConnection)
    }

    /// Enable MDS_DATA_EXPORT notifications and start streaming.
    ///
    /// Per the MDS spec the host must both:
    /// 1. Subscribe to CCCD notifications on the characteristic.
    /// 2. Write `0x01` (streaming enabled) to the characteristic value.
    open func enableDataExportNotification() async throws {
        // Reset sequence tracking for new stream
        lastSeqNumber = nil
        
        // Remove any stale notification registration so we always do a fresh CCCD write.
        removeCharacteristicNotification(BleMdsClient.MDS_DATA_EXPORT)
        // Re-register so notifyDescriptorWritten can update state and tearDown() works.
        automaticEnableNotificationsOnConnect(chr: BleMdsClient.MDS_DATA_EXPORT, disableOnDisconnect: true)
        // 1. Subscribe to CCCD — awaitCccdWrite issues the write and suspends until
        //    notifyDescriptorWritten resumes cccdContinuation with the ATT result.
        BleLogger.trace("BleMdsClient enabling CCCD for DATA_EXPORT")
        try await awaitCccdWrite(notify: true)
        // 2. Write mode byte 0x01 — awaits ATT Write Response via serviceDataWritten.
        BleLogger.trace("BleMdsClient enabling streaming (write 0x01 to DATA_EXPORT)")
        try await writeDataExportMode(0x01)
    }

    /// Stop streaming and unsubscribe from MDS_DATA_EXPORT notifications.
    ///
    /// Per the MDS spec the host must both:
    /// 1. Write `0x00` (streaming disabled) to the characteristic value.
    /// 2. Unsubscribe from CCCD notifications.
    open func disableDataExportNotification() async throws {
        // 1. Write mode byte 0x00 — awaits ATT Write Response via serviceDataWritten.
        BleLogger.trace("BleMdsClient disabling streaming (write 0x00 to DATA_EXPORT)")
        try await writeDataExportMode(0x00)
        // 2. Unsubscribe from CCCD only if we previously subscribed.
        if containsNotifyCharacteristic(BleMdsClient.MDS_DATA_EXPORT) {
            BleLogger.trace("BleMdsClient disabling CCCD for DATA_EXPORT")
            try await awaitCccdWrite(notify: false)
            removeCharacteristicNotification(BleMdsClient.MDS_DATA_EXPORT)
        }
    }

    /// Issue a CCCD enable/disable write for DATA_EXPORT and suspend until the ATT response
    /// arrives via `notifyDescriptorWritten(_:enabled:err:)`.
    ///
    /// Stores `cccdContinuation` BEFORE calling `setCharacteristicNotify` to handle mocks that
    /// fire `notifyDescriptorWritten` synchronously inside the call.
    private func awaitCccdWrite(notify: Bool) async throws {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            cccdContinuation = continuation
            guard let transport = gattServiceTransmitter else {
                cccdContinuation = nil
                continuation.resume(throwing: BleGattException.gattTransportNotAvailable)
                return
            }
            do {
                try transport.setCharacteristicNotify(
                    self,
                    serviceUuid: BleMdsClient.MDS_SERVICE,
                    characteristicUuid: BleMdsClient.MDS_DATA_EXPORT,
                    notify: notify
                )
                BleLogger.trace("BleMdsClient DATA_EXPORT CCCD notify=\(notify) write issued")
            } catch {
                cccdContinuation = nil
                continuation.resume(throwing: error)
            }
        }
    }

    /// Write a streaming-mode byte to the DATA_EXPORT characteristic (write-with-response).
    ///
    /// Registers a CheckedContinuation BEFORE issuing the write to avoid a race, then
    /// suspends until serviceDataWritten(_:err:) resumes it when the ATT response arrives.
    private func writeDataExportMode(_ mode: UInt8) async throws {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            // Store continuation first — serviceDataWritten may be called on any thread.
            writeModeContinuation = continuation
            guard let transport = gattServiceTransmitter else {
                writeModeContinuation = nil
                continuation.resume(throwing: BleGattException.gattTransportNotAvailable)
                return
            }
            do {
                try transport.transmitMessage(
                    self,
                    serviceUuid: BleMdsClient.MDS_SERVICE,
                    characteristicUuid: BleMdsClient.MDS_DATA_EXPORT,
                    packet: Data([mode]),
                    withResponse: true
                )
                BleLogger.trace("BleMdsClient DATA_EXPORT mode 0x\(String(mode, radix: 16)) write issued")
            } catch {
                writeModeContinuation = nil
                continuation.resume(throwing: error)
            }
        }
    }

    /// Validate chunk sequence number (byte 0, bits [4:0]) and filter duplicates.
    ///
    /// Per Memfault MDS spec, each chunk starts with a modulo-32 sequence counter in bits [4:0] of byte 0.
    /// This method detects and drops duplicate chunks, and logs warnings when chunks are dropped.
    ///
    /// - Parameter data: Raw chunk data from MDS_DATA_EXPORT notification
    /// - Returns: `true` if chunk should be yielded to stream, `false` if duplicate (should drop)
    private func validateAndFilterChunk(_ data: Data) -> Bool {
        guard data.count >= 1 else {
            BleLogger.error("MDS chunk too short (\(data.count) bytes), dropping")
            return false
        }
        
        // Extract sequence number from low 5 bits of byte 0 (modulo-32)
        let seqNum = data[0] & 0x1F
        
        // First chunk after stream start — accept any sequence number
        guard let last = lastSeqNumber else {
            BleLogger.trace("MDS first chunk seq=\(seqNum)")
            lastSeqNumber = seqNum
            return true
        }
        
        // Calculate expected next sequence number (modulo-32 wrap-around)
        let expected = (last + 1) % 32
        
        if seqNum == expected {
            // Normal sequential chunk
            lastSeqNumber = seqNum
            return true
        } else if seqNum == last {
            // Duplicate chunk — drop silently
            BleLogger.trace("MDS duplicate chunk seq=\(seqNum), dropping")
            return false
        } else {
            // Gap detected — some chunks were dropped by BLE layer
            let gap = (seqNum + 32 - expected) % 32
            BleLogger.error("MDS chunk gap: expected seq=\(expected), got=\(seqNum), ~\(gap) chunk(s) dropped")
            // Yield the chunk anyway — let Memfault cloud handle missing data
            lastSeqNumber = seqNum
            return true
        }
    }

    /// Called by the BLE stack when a write-with-response completes for DATA_EXPORT.
    override public func serviceDataWritten(_ chr: CBUUID, err: Int) {
        guard chr.isEqual(BleMdsClient.MDS_DATA_EXPORT) else { return }
        guard let continuation = writeModeContinuation else { return }
        writeModeContinuation = nil
        if err == 0 {
            continuation.resume()
        } else {
            continuation.resume(throwing: BleGattException.gattCharacteristicError)
        }
    }
}
